compiler/executable/match_/planner/vertex/constraint.rs

```diff
@@ -124,32 +124,34 @@ impl ConstraintVertex<'_> {
 
     pub(crate) fn direction_from_join_var(
         &self,
-        var: VariableVertexId,
-        include: &HashSet<VariableVertexId>,
-        exclude: &HashSet<VariableVertexId>,
-    ) -> Option<Direction> {
+        join_var: VariableVertexId,
+        _ongoing_step_produced: &HashSet<VariableVertexId>,
+        all_produced: &HashSet<VariableVertexId>,
+    ) -> Result<Direction, QueryPlanningError> {
+        debug_assert!({
+            let unbound_join_variables: Vec<VariableVertexId> = self
+                .variables()
+                .filter(|&var| self.can_join_on(var) && (!all_produced.contains(&var) || _ongoing_step_produced.contains(&var)))
+                .collect();
+            unbound_join_variables.len() != 0
+        });
         // First check if we are in a bound case, in which case we don't care about directions
-        match self {
-            Self::Links(_) | Self::Has(_) | Self::IndexedRelation(_) => {
-                let unbound_join_variables: Vec<VariableVertexId> = self
-                    .variables()
-                    .filter(|&var| self.can_join_on(var) && (!exclude.contains(&var) || include.contains(&var)))
-                    .collect();
-                if unbound_join_variables.len() < 2 {
-                    return None;
-                }
-            }
-            _ => {
-                return None;
-            }
-        }
-        // If unbounded, we choose direction based on the provided join variable
-        match self {
-            Self::Links(inner) => Some(Direction::canonical_if(inner.relation == var)),
-            Self::Has(inner) => Some(Direction::canonical_if(inner.owner == var)),
-            Self::IndexedRelation(inner) => Some(Direction::canonical_if(inner.player_1 == var)),
-            _ => None,
-        }
+        let (canonical_from, canonical_to) = match self {
+            Self::Links(links) => (links.relation, links.player),
+            Self::Has(has) => (has.owner, has.attribute),
+            Self::IndexedRelation(indexed) => (indexed.player_1, indexed.player_2),
+            _ => return Err(QueryPlanningError::UnimplementedJoinForConstraint {} ),
+        };
+        // It can't be a join var if it's being newly produced in this step
+        let direction = if join_var == canonical_from {
+            debug_assert!(all_produced.contains(&canonical_to) || !_ongoing_step_produced.contains(&canonical_to));
```

Comment:
If `all_produced` is a superset of `ongoing_step_produced`, this is a tautology. Should that have been `&&` as a negation of the filter above?

Reply:
Aren't all asserts meant to be tautologies? I'm updating the comment. The comment I've left is wrong though. I'm updating it - could you check if it makes sense?

Reply:
Is this assert meant to verify that `ongoing_step_produced` _is_ a subset of `all_produced`? There's a better way to phrase that.

Otherwise, this is saying that either `canonical_to` is in `all_produced`, or not in `ongoing_step_produced` (same as "not in `all_produced`").